### PR TITLE
Authenticate with vault via aws STS in iam mode

### PR DESCRIPTION
**Pull request by pyncc**
Broke this part out from vault-clj since it involves an awk sdk that would be too bulky for a core lib.

The authentication dispatch `:aws-iam` creates inputs for and calls to this vault auth endpoint: https://www.vaultproject.io/api/auth/aws/index.html#login

While possible to roll aws request signing by hand, reusing a library seemed prudent.

It creates an STS GetCallerIdentity request, signs it with iam credentials provided by either the default credentials chain (including instance profile metadata if present) or a set of test credentials useful for imitating another iam principal like an EC2 instance without having to run this code there.

This signed request is packaged up and sent to the vault aws auth endpoint where vault verifies the signature and identity with AWS as an authentication mechanism to vault itself.

I still need to add a documentation section on how to configure the aws auth backend, but this could follow after this PR.

### CODE REVIEW

File: dev/user.clj [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    [vault.client.ext.aws :as ext]
+    [vault.core :as vault])
+  (:import
+    com.amazonaws.DefaultRequest
+    com.amazonaws.Request
+    com.amazonaws.SignableRequest
+    com.amazonaws.auth.AWS4Signer
+    com.amazonaws.auth.AWSCredentials
+    com.amazonaws.auth.BasicSessionCredentials
+    com.amazonaws.auth.DefaultAWSCredentialsProviderChain
```
Comment (greglook) [inline dev/user.clj]: Style: these should be grouped by package. What do you think of enabling stile checks as one of the CI tasks?

File: project.clj [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  :description "Vault-clj extension to support aws iam instance authentication."
+  :url "https://github.com/amperity/vault-clj-aws"
+  :license {:name "Apache License"
+            :url "http://www.apache.org/licenses/LICENSE-2.0"}
+
+  :deploy-branches ["master"]
+  :pedantic? :abort
+
+  :dependencies
+  [[org.clojure/clojure "1.9.0"]
```
Comment (greglook) [inline project.clj]: Any reason not to use 1.10 here? We ought to be upgrading where possible I think.

File: src/vault/client/ext/aws.clj [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    (.addHeader "Content-Type" "application/x-www-form-urlencoded; charset=utf-8")
+    (.addHeader "Content-Length" "43")
+    (.setContent (-> payload (.getBytes "UTF-8") (io/input-stream)))))
+
+
+(defn credentials
+  "AWSCredentials to authenticate with STS. Will fall back to default chain including instance profile."
+  ^AWSCredentials
+  ([id secret token]
+   (BasicSessionCredentials. id secret token))
```
Comment (greglook) [inline src/vault/client/ext/aws.clj:51]: Seems like it would be good to also accept a 2-arg version for non-session credentials.

File: src/vault/client/ext/aws.clj [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+   (-> (DefaultAWSCredentialsProviderChain/getInstance)
+       (.getCredentials))))
+
+
+(defn signer
+  "Signing object used to perform AWS4 signing on sts request object."
+  ^AWS4Signer
+  []
+  (doto (AWS4Signer.)
+        (.setServiceName "sts")
```
Comment (greglook) [inline src/vault/client/ext/aws.clj]: Style: indent is off

File: src/vault/client/ext/aws.clj [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        (.setRegionName "us-east-1")))
```
Comment (greglook) [inline src/vault/client/ext/aws.clj]: Does this need to be configurable?

File: project.clj [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+   [org.clojure/tools.logging "0.4.0"]
+   [amperity/envoy "0.3.1"]
+   [amperity/vault-clj "0.6.6"]
+   [buddy/buddy-core "1.4.0"]
```
Comment (greglook) [inline project.clj]: Are you including this just for the Base64 code? I'd strongly recommend using Java's built in [`Base64`](https://docs.oracle.com/javase/8/docs/api/java/util/Base64.html) codec instead.

File: src/vault/client/ext/aws.clj [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+   "iam_request_headers" (-> (.getHeaders req)
+                             (json/encode)
+                             (b64/encode)
+                             (bytes->str))})
+
+
+(defmethod client/authenticate* :aws-iam
+  [client _ aws-ctx]
+  (let [{:keys [iam-role test-credentials]} aws-ctx
+        aws-creds ^AWSCredentials (or test-credentials (credentials))
```
Comment (greglook) [inline src/vault/client/ext/aws.clj]: Suggestion: instead of naming this `:test-credentials` how about accepting just `:credentials`? You could support a few different ways of providing this; I'm thinking something [along these lines](https://github.com/greglook/blocks-s3/blob/20413c59ae08ab684092f2bf4545bcd1a4b3a627/src/blocks/store/s3.clj#L96-L124).

**Review by greglook — CHANGES_REQUESTED** [review]

File: dev/user.clj (hunk 1 above)
Comment (pyncc) [reply dev/user.clj]: Yeah hadn't wired that stuff up yet but I will include style when I do.

File: project.clj (hunk 2 above)
Comment (pyncc) [reply project.clj]: sure y not

File: src/vault/client/ext/aws.clj (hunk 5 above)
Comment (pyncc) [reply src/vault/client/ext/aws.clj]: I don't think so -- the main endpoint "sts.amazonaws.com" used to be the only one and present in us-east-1 I guess? This specifically was from reading the vault cli go implementation where they mention detection doesn't work properly using the sdk.

File: project.clj (hunk 6 above)
Comment (pyncc) [reply project.clj]: nice, i'm not yet in the habit of remembering that lib.

File: src/vault/client/ext/aws.clj (hunk 7 above)
Comment (pyncc) [reply src/vault/client/ext/aws.clj]: i'm renaming the keyword for now and will make a note about supporting an alternative approach for a future change but for now for my use its YAGNI.

File: project.clj [hunk 8]
```diff
@@ -8,16 +8,17 @@
   :pedantic? :abort
 
   :dependencies
-  [[org.clojure/clojure "1.9.0"]
+  [[org.clojure/clojure "1.10.0"]
    [org.clojure/tools.logging "0.4.0"]
    [amperity/envoy "0.3.1"]
-   [amperity/vault-clj "0.6.6"]
-   [buddy/buddy-core "1.4.0"]
+   [amperity/vault-clj "0.6.6" :exclusions [commons-logging]]
```
Comment (greglook) [inline project.clj:14]: Should be `0.7.0` now?

File: src/vault/client/ext/aws.clj [hunk 9]
```diff
@@ -41,13 +44,15 @@
...
+(defn derive-credentials
   "AWSCredentials to authenticate with STS. Will fall back to default chain including instance profile."
   ^AWSCredentials
   ([id secret token]
    (BasicSessionCredentials. id secret token))
+  ([id secret]
+   (BasicAWSCredentials. id secret))
   ([]
-   (-> (DefaultAWSCredentialsProviderChain/getInstance)
+   (-> (DefaultAWSCredentialsProviderChain.)
```
Comment (greglook) [inline src/vault/client/ext/aws.clj:55]: Hmm, why this change? Maybe because we mess with the default singleton in our services? 🤔

File: src/vault/client/ext/aws.clj [hunk 10]
```diff
@@ -56,34 +61,39 @@
...
   []
   (doto (AWS4Signer.)
-        (.setServiceName "sts")
-        (.setRegionName "us-east-1")))
+    (.setServiceName "sts")
+    (.setRegionName "us-east-1")))
+
+
+(defn- str->b64str
+  [^java.util.Base64$Encoder b64 ^String s]
```
Comment (greglook) [inline src/vault/client/ext/aws.clj:69]: These are cheap enough to make (and you do few enough calls) that it would be cleaner to make a new one each call to this function.

File: src/vault/client/ext/aws.clj [hunk 11]
```diff
@@ -56,34 +61,39 @@
...
+  (-> s
+    (.getBytes "UTF-8")
+    (->> (.encodeToString b64))))
```
Comment (greglook) [inline src/vault/client/ext/aws.clj:72]: This threading form seems much less clear than a single line:
```clojure
(.encodeToString (Base64/getEncoder) (.getBytes s "UTF-8"))
```

**Review by greglook — CHANGES_REQUESTED** [review]
A few more thoughts, but close to :+1:
